matchtree.go

```diff
@@ -866,17 +872,39 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
 			return &noMatchTree{"const"}, nil
 		}
 	case *query.Language:
-		code, ok := d.metaData.LanguageMap[s.Language]
-		if !ok {
-			return &noMatchTree{"lang"}, nil
+		code, codeOk := d.metaData.LanguageMap[s.Language]
+		var mt matchTree
+
+		if !codeOk {
+			if d.metaData.IndexFeatureVersion < 12 {
+				// This is an old index file without more precise language mappings,
+				// fall back to generating file match patterns.
+				extsForLang := enry_data.ExtensionsByLanguage[s.Language]
+				if extsForLang != nil {
+					extFrags := make([]string, 0, len(extsForLang))
+					for _, ext := range extsForLang {
+						extFrags = append(extFrags, regexp.QuoteMeta(ext))
+					}
+					if len(extFrags) > 0 {
+						reMt := &regexpMatchTree{
```

Comment:
by directly creating the regexpMatchTree you don't get to take advantage of the trigram indexes. Is there not a way we can re-use that smarts we have here? eg return `d.newMatchTree(&query.Regexp{...})`.

How about this, in `indexData.simplify` you couple this logic by always returning true. Why not generate the equivalent `query.Regexp` in that bit of code and then you don't need to touch matchtree generation. The benefit of this is there is less coupling between simplify and newMatchTree

Reply:
done